diff --git a/testing/proxytest/proxytest.go b/testing/proxytest/proxytest.go
index 54d748e4be..08c51e379f 100644
--- a/testing/proxytest/proxytest.go
+++ b/testing/proxytest/proxytest.go
@@ -50,9 +50,10 @@ type Proxy struct {
 type Option func(o *options)
 
 type options struct {
-	addr        string
-	rewriteHost func(string) string
-	rewriteURL  func(u *url.URL)
+	addr          string
+	rewriteHost   func(string) string
+	rewriteURL    func(u *url.URL)
+	verifyRequest func(r *http.Request) error
 	// logFn if set will be used to log every request.
 	logFn           func(format string, a ...any)
 	verbose         bool
@@ -121,6 +122,14 @@ func WithRewriteFn(f func(u *url.URL)) Option {
 	}
 }
 
+// WithVerifyRequest calls f on the request before the proxy request is made.
+// If the verification returns an error, the proxy reutrns an HTTP 500 (for http proxy), or HTTP 502 (https)
+func WithVerifyRequest(f func(r *http.Request) error) Option {
+	return func(o *options) {
+		o.verifyRequest = f
+	}
+}
+
 // WithServerTLSConfig sets the TLS config for the server.
 func WithServerTLSConfig(tc *tls.Config) Option {
 	return func(o *options) {
@@ -199,10 +208,10 @@ func New(t *testing.T, optns ...Option) *Proxy {
 				requestID, w.statusCode, r.Method, r.URL, r.Proto, r.RemoteAddr))
 		}),
 	)
-	p.Server.Listener = l
+	p.Listener = l
 
 	if opts.serverTLSConfig != nil {
-		p.Server.TLS = opts.serverTLSConfig
+		p.TLS = opts.serverTLSConfig
 	}
 
 	u, err := url.Parse(p.URL)
@@ -288,6 +297,12 @@ func (p *Proxy) serveHTTP(w http.ResponseWriter, r *http.Request) {
 func (p *Proxy) processRequest(r *http.Request) (*http.Response, error) {
 	origURL := r.URL.String()
 
+	if p.opts.verifyRequest != nil {
+		if err := p.opts.verifyRequest(r); err != nil {
+			return nil, err
+		}
+	}
+
 	switch {
 	case p.opts.rewriteURL != nil:
 		p.opts.rewriteURL(r.URL)
@@ -312,6 +327,14 @@ func (p *Proxy) processRequest(r *http.Request) (*http.Response, error) {
 	// needed anyway, so remove it.
 	r.RequestURI = ""
 
+	// Add a Forwarded header
+	host, _, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		p.opts.logFn("[ERROR] could not parse remote address %q: %v", r.RemoteAddr, err)
+	} else {
+		r.Header.Add("Forwarded", "for="+host)
+	}
+
 	return p.client.Do(r)
 }
 
diff --git a/testing/proxytest/proxytest_test.go b/testing/proxytest/proxytest_test.go
index f367ced2b6..d4bebd017f 100644
--- a/testing/proxytest/proxytest_test.go
+++ b/testing/proxytest/proxytest_test.go
@@ -8,6 +8,7 @@ import (
 	"context"
 	"crypto/tls"
 	"crypto/x509"
+	"errors"
 	"io"
 	"net"
 	"net/http"
@@ -171,6 +172,56 @@ func TestProxy(t *testing.T) {
 				}
 			},
 		},
+		{
+			name: "Basic scenario, request verifier passes",
+			setup: setup{
+				fakeBackendServer:      createFakeBackendServer(),
+				generateTestHttpClient: nil,
+			},
+			proxyOptions: []Option{WithVerifyRequest(func(r *http.Request) error {
+				if r.Method == http.MethodGet {
+					return nil
+				}
+				return errors.New("unsupported method")
+			})},
+			proxyStartTLS: false,
+			request: testRequest{
+				method: http.MethodGet,
+				url:    "http://somehost:1234/some/path/here",
+				body:   nil,
+			},
+			wantErr: assert.NoError,
+			assertFunc: func(t *testing.T, proxy *Proxy, resp *http.Response) {
+				assert.Equal(t, http.StatusOK, resp.StatusCode)
+				if assert.NotEmpty(t, proxy.ProxiedRequests(), "proxy should have captured at least 1 request") {
+					assert.Contains(t, proxy.ProxiedRequests()[0], "/some/path/here")
+				}
+			},
+		},
+		{
+			name: "Basic scenario, request verifier fails",
+			setup: setup{
+				fakeBackendServer:      createFakeBackendServer(),
+				generateTestHttpClient: nil,
+			},
+			proxyOptions: []Option{WithVerifyRequest(func(r *http.Request) error {
+				if r.Method == http.MethodPost {
+					return nil
+				}
+				return errors.New("unsupported method")
+			})},
+			proxyStartTLS: false,
+			request: testRequest{
+				method: http.MethodGet,
+				url:    "http://somehost:1234/some/path/here",
+				body:   nil,
+			},
+			wantErr: assert.NoError,
+			assertFunc: func(t *testing.T, proxy *Proxy, resp *http.Response) {
+				assert.Equal(t, http.StatusInternalServerError, resp.StatusCode)
+				assert.Empty(t, proxy.ProxiedRequests(), "proxy does not capture unverified request")
+			},
+		},
 	}
 
 	for _, tt := range testcases {
@@ -379,7 +430,12 @@ func prepareMTLSProxyAndTargetServer(t *testing.T, targetHost string) (*Proxy, h
 
 func createFakeBackendServer() *httptest.Server {
 	handlerF := func(writer http.ResponseWriter, request *http.Request) {
-		// always return HTTP 200
+		if h := request.Header.Get("Forwarded"); h == "" {
+			// return 400 if the proxy has not added the Forwarded header.
+			writer.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		// return HTTP 200 otherwise
 		writer.WriteHeader(http.StatusOK)
 	}
 
